src/batch/src/executor/expand.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use risingwave_common::array::column::Column;
+use risingwave_common::array::{DataChunk, PrimitiveArray, Vis};
+use risingwave_common::catalog::{Field, Schema};
+use risingwave_common::error::ErrorCode::InternalError;
+use risingwave_common::error::{Result, RwError};
+use risingwave_common::types::DataType;
+use risingwave_common::util::chunk_coalesce::{DataChunkBuilder, SlicedDataChunk};
+use risingwave_pb::batch_plan::plan_node::NodeBody;
+
+use super::{BoxedDataChunkStream, BoxedExecutor, BoxedExecutorBuilder, Executor, ExecutorBuilder};
+use crate::task::BatchTaskContext;
+
+pub struct ExpandExecutor {
+    expanded_keys: Vec<Vec<usize>>,
+    child: BoxedExecutor,
+    schema: Schema,
+    identity: String,
+}
+
+impl Executor for ExpandExecutor {
+    fn schema(&self) -> &Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
+
+impl ExpandExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(self: Box<Self>) {
+        let mut data_chunk_builder = DataChunkBuilder::with_default_size(self.schema.data_types());
+
+        #[for_await]
+        for data_chunk in self.child.execute() {
+            // TODO: handle dummy chunk.
+            let data_chunk: DataChunk = data_chunk?.compact()?;
+            let cardinality = data_chunk.cardinality();
+            let (columns, vis) = data_chunk.into_parts();
+            assert_eq!(vis, Vis::Compact(cardinality));
+
+            let null_columns: Vec<Column> = columns
+                .iter()
+                .map(|column| {
+                    let array = column.array_ref();
+                    let mut builder = array.create_builder(cardinality)?;
+                    (0..cardinality).try_for_each(|_i| builder.append_null())?;
+                    Ok::<Column, RwError>(Column::new(Arc::new(builder.finish()?)))
```

Comment:
We can switch to a more efficient way to create a typed all-null column, maybe later.